### DISCUSSION TITLE
[2023-04-24] GUI Team Demo Day

### DISCUSSION THREAD

**Discussion by wdanilo** [enso-org/enso#6393] [Weekly Demo Day]
Last week I implemented several things:

1. The List Editor was almost* (see point 2) finished - it allows now for adding elements and its internal API was finished to deliver information about all changes made in GUI and via API.

https://user-images.githubusercontent.com/1623053/233996739-8d3b2785-7bed-47f8-ba5b-b9e4f052aeef.mov

2. I discovered that our slider widget has bugs and does not use modern EnsoGL functionality. I updated it to test the List Editor interaction with widgets. The one thing missing from List Editor is proper passing events to sub-widgets, like sliders. I will finish fixing slider and finishing this events handling in List Editor the next day (I'm almost there).

3. I fixed several small bugs in the application and I also spent a lot of time with different people on solving issues / planning the work.

**Comment by farmaazon** [comment]
Added a method to WASM file which drops all structures. All resources are freed, the visuals disappears and connections are closed.

![wasm-dropping-presentation](https://user-images.githubusercontent.com/3919101/234010377-b4ab2f76-b496-4c64-b389-9bee16036d40.gif)

While implementing I fixed leaks in several places (not noticed before, because we actually never dropped Graph Editor). I fixed them, and left in the codebase helper structures for tracking leaks. For example, if I see my FRP network is leaking (with all stuff captured), I may add `leak_detector::Trace` field to it:

```rust
#[derive(Clone, CloneRef, Debug)]
pub struct Network {
    data: Rc<NetworkData>,
    pub trace: leak_detector::Trace,
}
```
Then we enable a concrete network we want to check:
```rust
    frp.network().trace.enable();
```

In drop function we have a code checking if any `Trace` structure persisted dropping the entire application. If yes, the backtraces of their creation (both by direct instantiation or by cloning-ref existing object) are printed.
```rust
#[wasm_bindgen]
pub fn drop() {
    IDE.with(RefCell::take);
    EXECUTOR.with(RefCell::take);
    leak_detector::TRACKED_OBJECTS.with(|objects| {
        let objects = objects.borrow();
        if !objects.is_empty() {
            error!("Tracked objects leaked after dropping entire application!");
            error!("Leaked objects: {objects:#?}");
        }
    })
}
```

**Comment by JaroslavTulach** [comment]
Engine performance:
- #6336
    - Speeds up session/InitProtocolConnection request by ~300ms (~10%) 3051ms vs 2740ms
- #6335
- Looking into why the cascade of  `<|` functions is slow - #6384
    - Followup PoC: #6394
- Resolving issue where warnings are massively duplicated - #6372
- #6065

**Comment by vitvakatu** [comment]
Last week:

1. https://github.com/enso-org/enso/pull/6259 got merged after a bug fix on the engine side.
2. Read-only mode implementation finished with https://github.com/enso-org/enso/pull/6342, merged today.
3. Fixed the issue https://github.com/enso-org/enso/issues/6158 (and introduced https://github.com/enso-org/enso/issues/6385, which is currently WIP)
4. Investigated a bug https://github.com/enso-org/enso/issues/6228, this is still ongoing.

Read-only mode forbids certain actions in the GUI:

https://user-images.githubusercontent.com/6566674/232824737-bdd6f676-b20d-4218-af7d-4373e00cbe3e.mp4

**Comment by kazcw** [comment]
Completed:
- Fixed build failures caused by #6315.
- Fixed #6320, caused by logic bug from 2019.
- Fixed #6370 bug. Debugged a side issue, found needed backend functionality (#6374).

Current:
- #6125: I think this is caused by improper use of framebuffer/texture APIs.
  - I've found a couple issues with our API usage, but not what causes this.
  - I'm going to fix it by switching to the "Immutable" API for the persistent framebuffer--Khronos recommends using it because the "Mutable" API is so hard to use correctly.

Next:
- Continue the bug hunt.

**Comment by mwu-tow** [comment]
1. Replaced the code signing certificate for Windows. It expired last week, so needed to be replaced with the new one. Runners and CI configuration has been updated accordingly.
2. Finishing URL handling PR: code review fixes, further cross-platform testing, fixing the window focus issue. https://github.com/enso-org/enso/pull/6243 (there are still rough edges caused by https://github.com/enso-org/enso/issues/6360)
3. Fixing issue with project opening on macOS. https://github.com/enso-org/enso/issues/6362

**Reply by wdanilo** [reply]
4. PRs that are not up to date against develop but can be trivially merged, are mergable now, so you can merge them without waiting for CI again.

**Comment by Procrat** [comment]
1. I changed our project initialisation to show a spinner whenever we're in the process of opening a project (including creating a new one or switching). #6321

https://user-images.githubusercontent.com/607786/234019207-e34096c7-2cb3-4ae3-abd3-291fcaf43084.mp4

2. When the user is trying to rename the project to an invalid name, it now shows an error pop-up and stays in edit mode. #6366

https://user-images.githubusercontent.com/607786/234025360-6761f7d6-c13c-479c-a0ba-8af5d81f6e87.mp4

**Comment by Frizi** [comment]
Continued work on widgets on nodes. [The PR](https://github.com/enso-org/enso/pull/6347) has been reviewed and I'm in process of fixing requested changes.
- Fixed remaining issues with edge drawing and updates
- Implemented support for multiple widgets being attached to the same span tree node, used that to deduplicate text rendering code between label and dropdown
- Finished edge drag handling and implemented different hover area sizes for widgets at different level
- Discovered and fixed a performance regression caused by reintroduced eager initialization of dropdowns

![image](https://user-images.githubusercontent.com/919491/234032678-05c07fd2-5e24-4560-8652-a1b459b6d973.png)
